### ISSUE TITLE
[Proposal] Detect and mark inefficient ConcurrentDictionary usage

### ISSUE THREAD

**Issue by Therzok** [dotnet/roslyn-analyzers#2274]
#### Analyzer package

Either Microsoft.NetFramework.Analyzers or Microsoft.NetCore.Analyzers or both?

#### Diagnostic ID

Not sure, there's no existing one.

#### Repro steps

```
ConcurrentDictionary<T1, T2> dictionary;
dictionary.AddOrUpdate(key, value, (k, v) => value);
dictionary.AddOrUpdate(key, k1 => value, (k2, v2) => value);
```

`value` is the final return value in both the add, add factory and the update factory. When this pattern is detected, offer a code fix to change the code to:

```
dictionary[key] = value;
```

.NET core has another overload here:
https://source.dot.net/#System.Collections.Concurrent/System/Collections/Concurrent/ConcurrentDictionary.cs,a47a5058ee42983d

**Comment by Evangelink** [comment]
@mavasani Does it need to be moved to the `runtime` for evaluation?

**Comment by mavasani** [comment]
Yes, tagging @buyaa-n @carlossanlop. Can this be ported to dotnet/runtime and triaged?
